Add support for overriding S3 storage endpoint

johnparn opened the pull request:
The changes makes it possible to use RiakCS, a S3 compliant storage, instead of Amazon AWS S3. This may be handy for those who doesn't have the possibility to use AWS S3 or if there is a need to run a local storage as part of the development environment.

The changes are mostly about handling new configuration options, especially to override the endpoint of the S3 storage. Along with the code changes there is a short description how one can test RiakCS locally along with a sample configuration.

johnparn commented:
Hi!

What are the permissions supposed to look like for the registry? With RiakCS the following URL is used _http://example.com/<bucket>/<componentsDir>/_ for the components.

Using ´s3cmd´I can verify the files are present.

```
DIR                                         s3://foo/components/oc-client/
2018-01-22 08:27        95   s3://foo/components/components-details.json
2018-01-22 08:27        65   s3://foo/components/components.json
```

Trying to access _http://example.com/<bucket>/_ in a browser yields an access denied error. The same goes for _http://example.com/<bucket>/<componentsDir>/_ or any file i the dir. Some of these dirs/files have to be publicly accessible, right?

If necessary this can be remedied by setting the permissions of the individual files/dirs. For example this will make components.json publicly accessible but not the components dir. 

```
s3cmd -c s3.cfg --no-ssl setacl --acl-public  s3://foo/components/components.json
```

kmcrawford commented on packages/oc-s3-storage-adapter/index.js:
```diff
@@ -27,25 +27,52 @@ module.exports = function(conf) {
+  let awsConfig = new AWS.Config({
     accessKeyId: conf.key,
     secretAccessKey: conf.secret,
-    region: conf.region,
```
With the removal of region, won't this break existing users using the AWS s3?

johnparn replied:
Yes, you're right, let's add that back. All user's are not likely to use the default us-east-1.

nickbalestra commented on packages/oc-s3-storage-adapter/index.js:
```diff
@@ -27,25 +27,52 @@ module.exports = function(conf) {
     accessKeyId: conf.key,
     secretAccessKey: conf.secret,
-    region: conf.region,
+    signatureVersion: signatureVersion,
```
Perhaps we could use [Shorthand property names](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Object_initializer) here

johnparn replied:
Like this?

```
  const
    bucket = conf.bucket ? conf.bucket : '',
    sslEnabled = conf.sslEnabled === false ? false : true,
    s3ForcePathStyle = conf.s3ForcePathStyle ? true : false,
    signatureVersion = conf.signatureVersion ? conf.signatureVersion : AWS.Config.signatureVersion,
    httpOptions = { timeout: conf.timeout || 10000 };
```

nickbalestra replied:
Not really, not a big fun of coma separated assignements as it also get in the way when refactoring of working on it.

By `Shorthand property names` I mean:

```
const signatureVersion = ....
const s3ForcePathStyle = ...
const sslEnabled = ....


let awsConfig = new AWS.Config({
signatureVersion,
s3ForcePathStyle,
sslEnabled
)}
...
```

nickbalestra commented on packages/oc-s3-storage-adapter/index.js:
```diff
@@ -27,25 +27,52 @@ module.exports = function(conf) {
+  let awsConfig = new AWS.Config({
     accessKeyId: conf.key,
     secretAccessKey: conf.secret,
-    region: conf.region,
```
By removing region, will break existing users I guess?

nickbalestra commented on packages/oc-s3-storage-adapter/index.js:
```diff
@@ -27,25 +27,52 @@ module.exports = function(conf) {
     httpOptions.agent = conf.agentProxy;
   }
 
-  AWS.config.update({
```
Is there any further implication in moving from `config.update` to `Config` constructor?

johnparn replied:
It seems like creating an instance:

```
AWS.Config({sslEnabled: false})
console.log(aws);
```

gives the same result  as:

```
AWS.config.update({sslEnabled:false});
console.log(AWS.config);
```

If you used update before then perhaps we should stick with that.

nickbalestra replied:
I think is fine either way. I was just wondering if you were aware of any difference.

johnparn commented:
It would be much cleaner to keep the riak adapter separate. It is more less just the configuration handling that is the difference.

johnparn commented:
By the way, do have any specific idea of implementing the wrapper?

johnparn commented:
@nickbalestra Alright, I see. I was thinking about keeping the handling of the new config options in the S3-adapter as the new options are valid in a general S3 config and not Riak specific. It is just that Riak requires certain options to be set. However if the changes to handle new config options necessary for riak are to live in the new riak-adapter, then I can scrap this PR.

nickbalestra commented:
No worries, I think the PR is perfect and I think is good to have those overriding capabilities in the S3 adapter cause as you say, those are valid in a general s3 config. So my idea was to simply remove the `riak` specific documentation (readme) from this PR.
As a second PR once this get merged and  published, we can create an `oc-riak-storage-adapter`, where you require the `oc-s3-storage-adapter` with the new changes, and set what is needed in the there (thats what I mean for a wrapper), and we can then put the documentation there.
The final experience for the user should be cleaner as we'll set behind the scene everything that need to be set/overriden in the `oc-s3-storage adapter` specifically for riak so that the use doesn't have to mess too much around with it. Does this explain better the idea?

TLDR; I don't think we should get rid of this PR, we could simply take out the riak specific documentation, and perhaps add some tests for the new overrides/config, what you think?

johnparn commented:
Great, that means our minds are in sync! And I've implemented the suggestions that you made. Looking into the tests to see if I can fix them as well.

johnparn commented:
This new commit contains some of the changes that were requested. Added `region` back and converted to shorthand property names. 

@kmcrawford Regarding the tests I haven't really wrapped my head around how to test the endpoint properly. If you have any ideas how I may proceed, please let me know.

johnparn commented:
I've created the wrapper for riak in a separate dir, _oc-s3-riak-storage-adapter, and moved the documentation to the same dir. Will push that tomorrow.

nickbalestra commented on packages/oc-s3-riak-storage-adapter/package.json:
```diff
@@ -0,0 +1,28 @@
+{
+  "name": "oc-s3-riak-storage-adapter",
+  "version": "1.0.0",
```
Could you put the version to 0.0.1 so that is not yet a 1.0 release when we publish it?

nickbalestra commented on packages/oc-s3-riak-storage-adapter/README.md:
```diff
@@ -0,0 +1,131 @@
+# RiakCS as registry storage
+
```
Until this is stable and has proper test coverage we should add a note to mention to not use this in production yet, something like:
`⚠️ THIS ADAPTER IS IN ACTIVE DEVELOPMENT, DON'T USE IN PRODUCTION`

johnparn commented:
There one thing to improve. Instead of letting the user specify the endpoint like this:

```
endpoint: {
        protocol: 'http',
        hostname: 'localhost',
        port: '8080'
}  
```

it is better to just to write this:

```
      endpoint: 'http://localhost:8080' 
```

In that case the use can choose to omit for example the protocol 

```
      endpoint: 'localhost:8080' 
```

and in that case it will fall back to default https. 

Changing that also means less code for handling config.

Regarding the tests of the endpoint I had to a add a `getConfig` method to be able to get my hands on the actual config from the module. Perhaps not the cleanest way to do it but it works and I've created some tests to validate the endpoint settings.

nickbalestra commented:
Yes, we'll be doing something similar behind the scene for the `path` that is causing the issue you reported in #34.

Let's get this merged/published and then follow-up with smaller focused PRs to improve the riak storage. We should be on the safe side here as we'll be publishing the riak under 0.x and we have a big FLUX alert notice on its readme until we happy with it for others to rely on it.

Thanks again for contributing this @johnparn ❤️
